src/cinder/linux/GstPlayer.cpp

```diff
@@ -20,1298 +17,1183 @@
+    guard.unlock();
+
+    if( mFrontVBuffer ) {
+        mVideoTexture = ci::gl::Texture::create( mFrontVBuffer.get(), GL_RGBA, width(), height() );
+        if( mVideoTexture ) mVideoTexture->setTopDown();
+    }
 }
 
-void GstPlayer::createTextureFromMemory()
+void GstPlayer::createTextureFromID()
 {
-	mMutex.lock();
-	std::swap( mFrontVBuffer, mBackVBuffer );
-	mMutex.unlock();
+#if defined( CINDER_GST_HAS_GL )
+    std::unique_lock<std::mutex> guard( mMutex );
+    std::swap( mCurrentBuffer, mNewBuffer );
+    guard.unlock();
+
+    GLint textureID = getTextureID( mCurrentBuffer.get() );
 
-	if( mFrontVBuffer ) {
-		videoTexture = ci::gl::Texture::create( mFrontVBuffer, GL_RGBA, width(), height() );
-		if( videoTexture ) videoTexture->setTopDown();
-	}
+    if( textureID != 0 ) {
+        mVideoTexture = ci::gl::Texture::create( GL_TEXTURE_2D, textureID, mGstData.width, mGstData.height, true );
+        mVideoTexture->setTopDown();
+    }
+#endif
 }
 
-void GstPlayer::createTextureFromID()
+GLint GstPlayer::getTextureID( GstBuffer* newBuffer )
 {
-	// Pop any old buffers.
-	GAsyncQueue *mOutputVideoBuffers = nullptr;
-	mOutputVideoBuffers = (GAsyncQueue*)g_object_get_data( G_OBJECT(mGstAppSink), "queue_output_buf" );
-	GstBuffer* old = nullptr;
-	if( g_async_queue_length( mOutputVideoBuffers ) > 0 ) {
-		old = (GstBuffer*)g_async_queue_pop( mOutputVideoBuffers );
-	}
-	// We keep the buffer memory around until we are done with this texture.
-	auto deleter = [ old ] ( ci::gl::Texture *texture ) {
-			delete texture;
-			if( old ) gst_buffer_unref( old );
-	};
-
-	int currentTextureID = -1;
-	mMutex.lock();
-	currentTextureID = mGstTextureID;
-	mMutex.unlock();
-
-	videoTexture = ci::gl::Texture::create( GL_TEXTURE_2D, currentTextureID, width(), height(), true, deleter );
-	if( videoTexture ) {
-		videoTexture->setTopDown();
-	}
+    GLint id = 0;
+    if( ! newBuffer ) 
+        return id;
+#if defined( CINDER_GST_HAS_GL )
+    GstMemory *mem = gst_buffer_peek_memory( newBuffer, 0 );
+    g_assert( gst_is_gl_memory( mem ) );
+    id = ((GstGLMemory *) mem)->tex_id;
+#endif
+    return id;
 }
 
 ci::gl::Texture2dRef GstPlayer::getVideoTexture()
 {
-	if( mNewFrame ){
-		if( sUseGstGl ) {
-			createTextureFromID();
-		}
-		else { 
-			createTextureFromMemory();
-		}
-		mNewFrame = false;
-	}
-	return videoTexture;
+    if( mNewFrame ){
+        if( ! sUseGstGl ) {
+            createTextureFromMemory();
+        }
+        else {
+            createTextureFromID();
+        }
+        mNewFrame = false;
+    }
+    return mVideoTexture;
 }
 
 void GstPlayer::resetVideoBuffers()
 {
-	if( sUseGstGl ) {
-		resetGLBuffers();
-	}
-	else {
-		resetSystemMemoryBuffers();
-	}
+    if( sUseGstGl ) {
+        resetGLBuffers();
+    }
+    else {
+        resetSystemMemoryBuffers();
+    }
 }
 
 void GstPlayer::resetGLBuffers()
 {
-	if( mInputVideoBuffers ) {
-		while( g_async_queue_length( mInputVideoBuffers ) > 0 ) {

```

Comment:
Hmm, strange, I meant to comment on what's currently line 1131. That still has the unique_lock.